File: .leetcode/3.longest-substring-without-repeating-characters.go
package main

/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	var (
		dummy  map[byte]bool = make(map[byte]bool)
		left   int
		right  int
		maxLen int
		char   byte
		d      byte
	)
	for right < len(s) {
		char = s[right]
		dummy[char] = !dummy[char]
		right++
		for !dummy[char] {
			d = s[left]
			left++
			dummy[d] = !dummy[d]
		}
		if right-left > maxLen {
			maxLen = right - left
		}
	}

	return maxLen
}

// @lc code=end

File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python
#
# [3] Longest Substring Without Repeating Characters
#


# @lc code=start
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        self.asciiArr = [0] * 128
        self.left = 0
        self.right = 0
        self.maxLen = 0

        while self.right < len(s):
            c = ord(s[self.right])
            self.right += 1
            self.asciiArr[c] += 1

            while self.asciiArr[c] > 1:
                d = ord(s[self.left])
                self.left += 1
                self.asciiArr[d] -= 1

            if self.right - self.left > self.maxLen:
                self.maxLen = self.right - self.left

        return self.maxLen


# @lc code=end
